lib/psych/visitors/yaml_tree.rb

```diff
@@ -310,19 +310,20 @@ def visit_String o
           style = Nodes::Scalar::LITERAL
           plain = false
           quote = false
-        elsif o =~ /\n/
-          style = Nodes::Scalar::LITERAL
         elsif o == '<<'
           style = Nodes::Scalar::SINGLE_QUOTED
           tag   = 'tag:yaml.org,2002:str'
           plain = false
           quote = false
+        elsif o =~ /\n[^\Z]/  # match \n except blank line at the end of string
+          style = Nodes::Scalar::LITERAL
+        elsif o.length > line_width
```

Comment:
I'm slightly worried about the impact on performance here.  It adds a conditional and a hash lookup that could be cached.  Could you change the initializer to configure this instance with a lambda so that we don't have to call `line_width` for every string?

Something like this:

```ruby
test_lamba = if options[:line_width]
  width = options[:line_width]
  lambda { |o|  o.length > width }
else
  lambda { |_| false }
end

# Then in the elsif

elsif test_lambda.call(o)
```

(But with correct indentation and better variable names)